synapse/replication/tcp/handler.py

```diff
@@ -153,32 +154,62 @@ def start_replication(self, hs):
                     )
 
             # We've now caught up to position sent to us, notify handler.
-            await self.replication_data_handler.on_position(cmd.stream_name, cmd.token)
+            await self._replication_data_handler.on_position(cmd.stream_name, cmd.token)
 
-        self.streams_connecting.discard(cmd.stream_name)
+        self._streams_connected.add(cmd.stream_name)
```

Comment:
this looks racy: if we get two `positions` each of which trigger reconnect/catchup, we'll add to `_streams_connected` while the second is still catching up.

Reply:
This should be in the linearizer